### DISCUSSION TITLE
关于复现Wisckey的想法

### DISCUSSION THREAD

**Discussion by WangTingZheng** [WangTingZheng/Paperdb#12] [Ideas]
# 主要步骤
主要分为以下五步
- Step1：构建基础的KV分离
     - 构建SSTable时，先把Key-Value写入vLog，获得Handle后把Handle作为Value写入SSTable
     - 读取SSTable时，只修改Block的读取逻辑，改造[DecodeEntry](https://github.com/WangTingZheng/Paperdb/blob/068d5ee1a3ac40dabd00d211d5013af44be55bea/table/block.cc#L55C27-L55C38)函数，在读取Value，也就是Handle后，从vLog读取数据再返回
- Step2：实现fallocate，文件打孔操作，可以释放一定区间的数据，为GC做准备
     -  LevelDB抽象了Env，所有跨平台，包括文件操作都放在这里
     -  LevelDB提供了了Env的单元测试
- Step3：为vLog实现GC
     -  遍历vLog，解析出key，到LSM Tree中查询，如果不存在，就用打孔操作释放掉
- Step4：用多线程加速随机读
     -  Wisckey[1]的vLog的性能瓶颈在vLog，主要是随机读慢
     - SSD的随机读可以用多线程加速
- Step5：更细粒度的KV分离
     - 类似DiffKV[2]或者Titan[3]的思路
     - 大的放vLog，小的放LSM Tree
     - 可能要在InternalKey中再加一个flag
# 参考文献

[1] Lu L, Pillai T S, Gopalakrishnan H, et al. Wisckey: Separating keys from values in ssd-conscious storage[J]. ACM Transactions on Storage (TOS), 2017, 13(1): 1-28.

[2] Li Y, Liu Z, Lee P P C, et al. Differentiated Key-Value Storage Management for Balanced I/O Performance[C]//USENIX Annual Technical Conference. 2021: 673-687.

[3] PingCAP. Titan: A Distributed Redis Protocol Compatible NoSQL Database. https://docs.pingcap.com/tidb/dev/titan-overview.

**Comment by WangTingZheng** [comment]
更新：直接改造Block实现KV分离的读取有一些缺陷，主要有以下两点原因：
- 并不是所有的Block都需要KV分离，比如说Index Block/Meta Index Block就不用，不过Step5是要支持细粒度的KV分离的
- 读取KV分离的Value需要有文件句柄，这个需要从DB一直传到Block，要破坏很多文件的接口，不过句柄可以存在Option里？

但是Block级别的KV分离有好处：
- LevelDB的范围查询和点查询都是基于Block的迭代器的，Block支持KV分离，实际上就使得范围查询和点查询同时支持了，上层的接口不需要改动

**Reply by WangTingZheng** [reply]
目前用了一个更好的方法，在Block的迭代器外层套一层迭代器，这个迭代器会把Block的迭代器的value当作Handle，去vlog中读取真正的value，这么做有两个好处：
- 几乎不会破坏block的实现，只需要传入几个参数即可
- 可以从根上支持KV分离，LevelDB的整个磁盘的部分的查询，不管是点查询还是范围查询，都是依赖于block的迭代器的
但是也有一些问题
- Value()函数是const的，而解析handle读取value的过程中可能会出现一些错误，所以需要错误提示，这里会破坏const的特性

**Comment by WangTingZheng** [comment]
目前从CI中发现的一些问题：
- clang相比gcc更加严格，比如说这里有一个[switch分支](https://github.com/WangTingZheng/Paperdb/blob/e0d8fa1485809833dc4890e0c7e193a41c113de9/db/db_impl.cc#L246C7-L246C20)来分类不同的文件类型，我们增加了Vlog类型，但是并没有在这个分支增加它，gcc不会报错，而clang会报错，可看这个[commit](https://github.com/WangTingZheng/Paperdb/commit/e0d8fa1485809833dc4890e0c7e193a41c113de9)。里面可以看到Github Action的报错。
- 在macos平台下，参数是uint64的类型的指针，并不能传入一个size_t变量的地址，其它平台是可以的，这可能是和size_t的跨平台实现有关系，可以看这个[commit](https://github.com/WangTingZheng/Paperdb/commit/79091aa93e4219117a336b070be4997112f84d83)。

**Comment by WangTingZheng** [comment]
用Clion可能会出现SAN无法检测出内存泄漏的情况，主要的原因是Clion的设置下的Dynamic Analysis Tools下的Sanitizers里的AddressSanitizer里加了一个不检测内存泄漏的参数。

而且Leak Sanitizers不能和Debug模式一起用，只能Run程序。

**Comment by WangTingZheng** [comment]
各个平台的路径不一样，不能直接硬编码类似``/tmp/db``这样的路径，需要用``GetTestDirectory()``函数来获得当前系统的路径，不然Windows系统会找不到路径。

**Comment by WangTingZheng** [comment]
Block level的KV分离不太行，要改的地方太多了，Compaction内部都要改，还是DB level的KV分离比较好，只需要抓住两个地方：
- 入口：buildTable是数据要进入磁盘的唯一途径
- 出口：db_->NewIterator()：点查和范围查询都基于它

**Comment by KKould** [comment]
WiscKey已经是大改整体结构了，这样其他论文的复现执行起来就很困难了吧

**Reply by WangTingZheng** [reply]
主要是很多论文都依赖它，都用了KV分离，你要是找那些对整体结构改动小的，也能找到，LSM Tree的论文很多很多。。。

**Reply by WangTingZheng** [reply]
SSTable->Cache->Memtable->Compaction->MVCC->崩溃恢复，按这个路径，选改动最小的，一个个复现，到后面就熟悉了，就可以大刀阔斧地改了，精读论文/复现论文对对LevelDB的认识的提升是巨大的，因为写论文不仅仅要东西硬，写得得好，不复现论文光靠自己想，点子基本上都是不靠谱的。。
